src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java

```diff
@@ -0,0 +1,364 @@
+			}
+			else
+			{
+				logger.error("File {} is not readable!", pathToFile);
+			}
+		}
+		else
+		{
+			// If something failed during the data retrieval, the file might not exist. If that happens, let the user know.
+			logger.error("File \"{}\" still does not exist after executing the file retriever!", pathToFile);
+		}
+	}
+
+	protected void downloadData() throws Exception
+	{
+		logger.trace("Scheme is: {}", this.uri.getScheme());
+		Path path = null;
+		try
+		{
+			path = Paths.get(this.destination);
+			Files.createDirectories(path.getParent());
+			if (this.uri.getScheme().equals("http") || this.uri.getScheme().equals("https"))
+			{
+				
+				doHttpDownload(path);
+			}
+			else if (this.uri.getScheme().equals("ftp") || this.uri.getScheme().equals("sftp"))
+			{
+				doFtpDownload();
+			}
+			else
+			{
+				throw new UnsupportedSchemeException("URI "+this.uri.toString()+" uses an unsupported scheme: "+this.uri.getScheme());
+			}
+		}
+		catch (URISyntaxException e)
+		{
+			logger.error("Error creating for download destination: " + this.destination, e);
+			e.printStackTrace();
+		}
+		catch (IOException e)
+		{
+			logger.error("Unable to create parent directory of download destination: " + path.toString(), e);
+			e.printStackTrace();
+		}
+		catch (Exception e)
+		{
+			logger.error("Error performing download!", e);
+			throw e;
+		}
+		
+
+	}
+
+	protected void doFtpDownload() throws SocketException, IOException, FileNotFoundException, Exception
+	{
+		doFtpDownload(null, null);
+	}
+	
+	protected void doFtpDownload(String user, String password) throws SocketException, IOException, FileNotFoundException, Exception
+	{
+		user = user == null || user.trim().equals("") ? "anonymous" : user;
+		password = password == null || password.trim().equals("") ? "" : password;
+		FTPClient client = new FTPClient();
+		
+		client.connect(this.uri.getHost());
+		if (this.passiveFTP)
+		{
+			client.enterLocalPassiveMode(); //PASSIVE mode works better when inside a docker container.
+		}
+		client.login(user, password);
+		logger.debug("connect/login reply code: {}",client.getReplyCode());
+		client.setFileType(FTP.BINARY_FILE_TYPE);
+		client.setFileTransferMode(FTP.COMPRESSED_TRANSFER_MODE);
+		try
+		{
+			try(InputStream inStream = client.retrieveFileStream(this.uri.getPath()))
+			{
+				if (inStream != null)
+				{
+					writeInputStreamToFile(inStream);
+				}
+				else
+				{
+					logger.error("No data returned from server for {}", this.uri.toString());
+				}
+			}
+		}
+		catch (IOException e)
+		{
+			logger.error("Error while retrieving the file: {}",e.getMessage());
+			e.printStackTrace();
+			throw new IOException(e);
+		}
+		//Should probably have more/better reply-code checks.
+		logger.debug("retreive file reply code: {}",client.getReplyCode());
+		if (client.getReplyString().matches("^5\\d\\d.*") || (client.getReplyCode() >= 500 && client.getReplyCode() < 600) )
```

Comment:
Why the regex in addition to the check of between 500 and 600?

Reply:
The regex is matches against replyString, the range is matched against replyCode. I imagine there was some situation where a server sent an error in one field but not the other, but this code is rather old so I can't recall the _exact_ situation.